### PR TITLE
Separate car/track routing

### LINKED ISSUE #1053: trams are unable to turn based on lane arrows on shared car-tram lanes.
### Describe the problem
TMPE appears to apply overly strict restrictions on trams changing lanes on junctions, preventing tram lines from be placed in situations where it shouldn't.

In this example I've got a 4-lane 2-way road with tram tracks running down the middle (https://steamcommunity.com/sharedfiles/filedetails/?id=2130577793) and the Snowfall DLC 2-lane 1-way road with tram track:

I am placing a tram stop on the 4-lane 2-way road (on the right), and then attempting to place the next stop on the 2-lane road. As you can see, the game won't let me to:

I am guessing TMPE is forbidding all vehicles using the left lane on the 4-lane road from entering the right lane on the 2-lane road. While this makes sense for cars, this restriction should not be applied to trams.

The issue does *not* occur if the 4-lane road is replaced with a similar road, where the central lanes are tram only (https://steamcommunity.com/sharedfiles/filedetails/?id=2130576379).

The issue only occurs when using TMPE.

### Steps to reproduce
1.  Build a junction like on the screenshot using the 4L 2-way road with tram tracks (https://steamcommunity.com/sharedfiles/filedetails/?id=2130577793) and the Snowfall DLC 2L 1-way road with tram track.
2. Place a tram stop on the 4L road.
3. Attempt to place the next tram stop on the 2L road. Notice that the game will refuse to connect them.

### PR DESCRIPTION
**Pull request by kianzarrin**
[TMPE.zip](https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=1053-mixed-lanes)

- Added a car/track flag to routing.
- Separated the code path for car/track in routing manager for simplicity.
- routing manager does not route tracks at acute turns.

minor changes:
- track lane connection does not get mixed in the same array as car lane connection.
- `outgoingVehicleLanes` and `incommingVehicleLanes` now only applies to cars lanes only. Also renamed the variable to `incommingCarLanes\incommingVehicleLanes`

This serves as a precursor to #354. 
also fixes #1053. Test:
<div><image width="256px" src="https://user-images.githubusercontent.com/26344691/160487526-7cfe1849-13d1-467d-b8a1-37d780bc8788.png" /></div>

### CODE REVIEW

File: TLM/TLM/Manager/Impl/RoutingManager.cs [hunk 1]
```diff
@@ -581,8 +582,8 @@ public class RoutingManager
                     }
                 }
 
-                int outgoingVehicleLanes = 0;
-                int incomingVehicleLanes = 0;
+                int outgoingCarLanes = 0;
+                int incomingVCarLanes = 0;
```
Comment (aubergine10) [inline TLM/TLM/Manager/Impl/RoutingManager.cs:586]: `VCar`? typo?

File: TLM/TLM/Manager/Impl/RoutingManager.cs [hunk 2]
```diff
@@ -733,319 +732,356 @@ public class RoutingManager
...
+                                                    "{2}, {3}): nextLaneId={4}, idx={5} has outgoing connections " +
+                                                    "but is NOT connected with previous lane",
+                                                    prevSegmentId,
+                                                    prevLaneIndex,
+                                                    prevLaneId,
+                                                    isNodeStartNodeOfPrevSegment,
+                                                    nextLaneId,
+                                                    nextLaneIndex);
+                                            }
+                                        }
```
Comment (aubergine10) [inline TLM/TLM/Manager/Impl/RoutingManager.cs:795]: `else if (extendedLogRouting) ...` ?

File: TLM/TLM/Manager/Impl/RoutingManager.cs [hunk 3]
```diff
@@ -733,319 +732,356 @@ public class RoutingManager
...
+
+                                        bool connected;
+                                        if (stayInlane) {
+                                            if (extendedLogRouting) {
+                                                Log._Debug($"similar track networks: Preventing lane changes.");
+                                            }
+                                            connected = outerSimilarLaneIndexMatches;
+                                        } else {
+                                            connected = goodTurnAngle;
+                                        }
```
Comment (aubergine10) [inline TLM/TLM/Manager/Impl/RoutingManager.cs:813]:
```csharp
connected = stayInLane
    ? outerSimilarLaneIndexMatches
    : goodTurnAngle;

Log._DebugIf(extendedLogRouting && stayInLane, ...); // auto-removed from non-DEBUG builds
```

File: TLM/TLM/Util/TrackUtils.cs [hunk 4]
```diff
@@ -53,5 +57,39 @@ internal static class TrackUtils {
...
+
+            return true;
+        }
+
+        public static LaneEndTransitionGroup GetLaneEndTransitionGroup(VehicleInfo.VehicleType vehicleType) {
+            LaneEndTransitionGroup ret = 0;
+            if (vehicleType.IsFlagSet(LaneArrowManager.VEHICLE_TYPES))
+                ret |= LaneEndTransitionGroup.Car;
+            if (vehicleType.IsFlagSet(TrackUtils.VEHICLE_TYPES))
+                ret |= LaneEndTransitionGroup.Car;
```
Comment (aubergine10) [inline TLM/TLM/Util/TrackUtils.cs:91]: Should this be `Track`?

**Comment by aubergine10** [comment]
@krzychu124 how come this PR is mergable without any reviews? 
image

File: TLM/TLM/Util/TrackUtils.cs (hunk 4 above)
Comment (krzychu124) [reply TLM/TLM/Util/TrackUtils.cs:91]: Yeah I also think it's weird the const was defined there

File: TLM/TLM/Manager/Impl/RoutingManager.cs (hunk 3 above)
Comment (kianzarrin) [reply TLM/TLM/Manager/Impl/RoutingManager.cs:813]: `extendedLogRouting` is `const` in release builds. so `if(extendedLogRouting)` will be ignored in release builds.

**Comment by aubergine10** [comment]
build fail

**Comment by kianzarrin** [comment]
oops ... extra bracket

File: TLM/TLM/Util/TrackUtils.cs (hunk 4 above)
Comment (aubergine10) [reply TLM/TLM/Util/TrackUtils.cs:91]: @kianzarrin are tracked lanes supposed to be setting `LaneEndTransitionGroup.Car` here?

Comment (kianzarrin) [reply TLM/TLM/Util/TrackUtils.cs:91]: Yes. this is a bug. Will confuse VehicleAI.

Comment (kianzarrin) [reply TLM/TLM/Util/TrackUtils.cs]: OK I fixed that but now 1053 is not fixed anymore.

Comment (kianzarrin) [reply TLM/TLM/Util/TrackUtils.cs]: maybe before it appeared fixed because it was routing for cars

**Comment by kianzarrin** [comment]
Now it passes test. its ready to review.

**Comment by kianzarrin** [comment]
@aubergine10 do you understand enough of this code to be able to approve it? 
the routing detector PRs are based on this so you can use them to see the routings.

**Comment by kianzarrin** [comment]
@WildQuoll can you please test this. you can use the TMPE.zip link at the top of the page.
image

if you stress test it in your tram city it would be nice (don't save just test).

**Comment by kianzarrin** [comment]
image
when I change lane connections tram sometimes gets stuck.
I didn't add lane connection to the tram lane though. I think what triggered the problem was publish segment changes.
Routing does exist though:
image
Is this know issue?

**Comment by kianzarrin** [comment]
@aubergine10 what is the plan for this PR? when will it be reviewed?

**Review by kvakvs — APPROVED** [review]
Code looking fine

**Review by krzychu124 — CHANGES_REQUESTED** [review]
Routing for trolleybuses is completely broken.

https://user-images.githubusercontent.com/19638970/164845142-f0765fb0-4e1d-4fcc-897e-9a2158d1d489.mp4

**Comment by aubergine10** [comment]
Code looks good, but haven't had chance to test in game yet. Did you create issue for that tram sinking issue you found?

**Comment by kianzarrin** [comment]
@krzychu124 trolleybuses are not track or lane-arrow-following vehicle. which one should they be?

**Comment by aubergine10** [comment]
@kianzarrin I think we should treat trolleybuses as following "overhead tracks" (the overhead wires). That being the case, they should be able to ingore lane arrows. If user wants more control over the "overhead tracks", maybe we could do something with lane connectors?

As for trams, similar thing - they should be able to go wherever the in-road tracks allow them to go. And if player wants to control the connections of tracks at junctions use lane connectors.

I think same with buses - if there's a bus lane the bus should be able to use it, regardless of lane arrows (obviously lane direction is still obeyed), otherwise why is the bus lane there? It's only on normal road lanes that buses should be following lane arrows, I think.

**Comment by krzychu124** [comment]
> I think we should treat trolleybuses as following "overhead tracks" (the overhead wires). That being the case, they should be able to ingore lane arrows. If user wants more control over the "overhead tracks", maybe we could do something with lane connectors?

I also think they should ignore lane arrows. Trolleybuses actually don't care if there are overhead wires or not. They only require a lane with `Trolleybus` vehicle type so they work more or less like trams but unlike trams they should be able to ignore turn angle just like other road vehicles.

**Comment by kianzarrin** [comment]
@aubergine10 @krzychu124 
Note that Trollybuses can use any of the road lanes regardless of the position of the "overhead tracks".
currently there is a option about "buses can ignore lane arrows". I think that option should influence trollybuses too.

In the image bellow I think both bus and trolley bus should use the left-most lane to turn left:
image

But here I think they can use either:
image

Do you agree?

**Comment by krzychu124** [comment]
> Note that Trollybuses can use any of the road lanes regardless of the position of the "overhead tracks".

@kianzarrin That's not true -> oneway trolleybus road
image

They should always ignore arrows. Just try Dedicated turn lanes feature and try creating route on one-way trolleybus roads (left/right will become only right on screenshot above).

Also.. pathfinding **still does not work**. I can create trolleybus line but vehicles cannot be spawned (flicker 0/1),

I would make trolleybuses to completely ignore lane arrows. Later when Lane connector will get option for tram connections it could support trolleybuses if such lane type was selected.

**Comment by kianzarrin** [comment]
> Also.. pathfinding still does not work. I can create trolleybus line but vehicles cannot be spawned (flicker 0/1),

it works now. but I am yet to test cargo ships and ferries.

**Comment by kianzarrin** [comment]
@krzychu124 are ships/ferrys/planes even handled by CustomPathFind?

**Comment by krzychu124** [comment]
> I made them ignore lane arrows.

Does not work... 😞 

https://user-images.githubusercontent.com/19638970/164943677-4ad8eac1-ab5e-47ab-b90d-10ca7c76f924.mp4

> are ships/ferrys/planes even handled by CustomPathFind?

Everything. If we don't patch their StartPathFind then they just don't use any TM:PE features so don't touch Pathfinding if not needed (in 99% cases it is not required).

**Review by krzychu124 — APPROVED** [review]
OK, works, but only if you pull branch again or use zip file _(do not force push changes)_

**Comment by kianzarrin** [comment]
Tested with this city too: https://steamcommunity.com/sharedfiles/filedetails/?id=2561345715
It has everything but trolleybus.
Screenshot (47)

**Comment by aubergine10** [comment]
Is this PR suitable for inclusion in to 11.6.5.2?

**Comment by aubergine10** [comment]
Bit late, but finally got round to testing this in game and didn't encounter any problems so LGTM 👍

**Pull request merged**